### ISSUE TITLE
Custom metric: top three secondary pages

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/httparchive.org#400]
Feature request: create a custom metric that uses heuristics to guess what the top three URLs are for secondary pages linked from the home page. In other words, which pages would users be most likely to navigate to next?

This is part of an early exploration into how we might be able to test pages other the home page. Our source dataset is the Chrome UX Report, which gives us the top origins used by real Chrome users. Right now we simply test the root page for each origin, but the problem with this approach is that landing pages like the home page are commonly built differently from interior/secondary pages on the site. So if we had a way to test those additional pages, we could improve the representativeness of the HTTP Archive dataset. For example, React may be less likely to be used on a landing page, so our methodology may be causing us to undercount websites that use React.

We don't currently have capacity to start testing more than one page per origin, but this custom metric can be used to manually validate and fine-tune the approach until the capacity issue is resolved.

Let's use this issue to brainstorm an ordered list of potential heuristics.

**Comment by rviscomi** [comment]
@pmeenan suggested looking at the LCP element and taking its ancestor's `a[href]`, if it exists. This seems like it would be a very strong signal.

Another approach is to query the DOM for the first anchor elements. We could prioritize the search to things like `main` or `header` ancestors, for example: `document.querySelectorAll('main a, header a, a')`. This might help avoid low-quality pages like the footer TOS and privacy policy links.

**Comment by brendankenny** [comment]
Yes! Would be very excited to see a HTTP Archive URLs + 1 hop dataset.

- > Same origin

  I was going to ask about this. It makes sense to keep things contained.
   - Would this be same origin as the page, _after all redirects_?
   - would it make sense to relax this for the data collection stage? Wouldn't want to flood the dataset with irrelevant links, but it might be useful for sites with no (or barely any) prominent same-origin links on the page to still collect links and see if there's any relation worth exploring (subdomains or whatever)
 
- > Normalized (drop query string params and hash)

   this could drop marketing params useless for HTTP Archive, but could sometimes be vital for page content?

- Would it be feasible to check if the URL has CrUX data associated with it? I guess we could also do this outside of the run once we have a list of candidates. Same origin, prominent on the page, _and_ has URL-level CrUX data would be a powerful combo :)

**Comment by patrickhulce** [comment]
I love this idea and direction! 🎉 

My rough proposal for the 3 to balance importance + diversity:

- LCP a[href] OR anchor with the largest bounding rect / font-size in the viewport (capturing primary call to action)
- anchor with the largest bounding rect / font-size out of viewport (capturing secondary calls to action)
- randomly sampled anchor (not biasing the set *too* much, ensuring diversity)

**Comment by rviscomi** [comment]
Thanks for the input @brendankenny and @patrickhulce!

> Would this be same origin as the page, after all redirects?

> Would it be feasible to check if the URL has CrUX data associated with it? I guess we could also do this outside of the run once we have a list of candidates. Same origin, prominent on the page, and has URL-level CrUX data would be a powerful combo :)

My sense is that we should decide at runtime in the custom metric itself whether the URL is viable or not. So we won't have the chance to probe the URL to see where it redirects. Similarly for the CrUX idea, I don't think it's feasible to check for URL-level data at runtime.

> would it make sense to relax this for the data collection stage? Wouldn't want to flood the dataset with irrelevant links, but it might be useful for sites with no (or barely any) prominent same-origin links on the page to still collect links and see if there's any relation worth exploring (subdomains or whatever)

I think that's a good idea as a fallback strategy to ensure that we have some kind of interior coverage for pages that would otherwise be filtered out as low priority. For example, if there are no same-origin links, maybe we can settle for same-TLD+1 (different subdomain).

> this could drop marketing params useless for HTTP Archive, but could sometimes be vital for page content?

We could compare the error rates of pages with and without normalization to get a better sense of how big of an issue it is.

> My rough proposal for the 3 to balance importance + diversity:
>
>- LCP a[href] OR anchor with the largest bounding rect / font-size in the viewport (capturing primary call to action)
>- anchor with the largest bounding rect / font-size out of viewport (capturing secondary calls to action)
>- randomly sampled anchor (not biasing the set too much, ensuring diversity)

+1 that sounds like a great starting point.

**Comment by bkardell** [comment]
Great to see this. I *literally* just asked Rick abuot the status of this because I was in a meeting two years ago where this was discussed and IIRC I was very much of the opinion that while it is great we expand the number of pages in the dataset (I think it is ~2x what it was then), there are very diminishing returns to the value compared to collecting additional pages equivalent to similar growth (ie, if that is right - 2 pages rather than one) - and that even a random same origin page would give us far better insight.

I support something as simple as that, or any potential improvement that more carefully selects an additional page.

**Comment by rockeynebhwani** [comment]
@rviscomi - Another heuristics to find search pages on any site is to look for OpenSearch tag (https://en.wikipedia.org/wiki/OpenSearch)

Example - https://www.myntra.com/

You will find an entry like this - <link rel="search" type="application/opensearchdescription+xml" href="https://www.myntra.com/opensearch.xml" title="Myntra Fashion Search">

and from XML file, you can pick URL - https://www.myntra.com/{searchTerms}?userQuery=true

We can replace {searchTerms} with any random term and this will hit Search page on that page site. Even if we don't replace it with a random term and just hit it as-is, it will execute a search.

I am not sure how widely this tag is used but this will be a way to find Search pages on a site.

**Comment by philipwalton** [comment]
> @pmeenan suggested looking at the LCP element and taking its ancestor's `a[href]`, if it exists. This seems like it would be a very strong signal.

One comment with this heuristic: for several of the sites I work on, I wrap content images in an `<a>` element that links to a higher res version of the image. I don't know how common that is, but it's probably worth falling back to something else if the content type from that response is not HTML.

**Comment by tunetheweb** [comment]
> One comment with this heuristic: for several of the sites I work on, I wrap content images in an `<a>` element that links to a higher res version of the image. I don't know how common that is, but it's probably worth falling back to something else if the content type from that response is not HTML.

Good point but in most cases doesn't that open the image in separate window (e.g. https://example.com/images/example.jpg)? So we could exclude URLs ending in common image extensions - though admittedly not everyone uses image extensions.

Probably a good filter to apply no matter what the criteria so avoid "wasting" secondary "pages" on images and other media anyway. Maybe even if we see a media `content-type` on response we should through it away and try the next candidate?

**Comment by rockeynebhwani** [comment]
> @rviscomi - Another heuristics to find search pages on any site is to look for OpenSearch tag (https://en.wikipedia.org/wiki/OpenSearch)
> 
> Example - https://www.myntra.com/
> 
> You will find an entry like this - `<link rel="search" type="application/opensearchdescription+xml" href="https://www.myntra.com/opensearch.xml" title="Myntra Fashion Search">`
> 
> and from XML file, you can pick URL - `https://www.myntra.com/{searchTerms}?userQuery=true`
> 
> We can replace {searchTerms} with any random term and this will hit Search page on that page site. Even if we don't replace it with a random term and just hit it as-is, it will execute a search.
> 
> I am not sure how widely this tag is used but this will be a way to find Search pages on a site.

Just expanding on this.... 

Identification of search pages can be improved further by looking for Schema.org tags .. We will have to look for @type=SearchAction under potentialAction. Example below from www.flipkart.com. We have some stats in last year's SEO chapter on usage of this.. see here - https://almanac.httparchive.org/en/2020/seo#most-prevalent-structured-data-objects

![image](https://user-images.githubusercontent.com/1718757/132852284-7297311c-b7dd-4db8-bc42-24ae4ad4faa2.png)

**Comment by tunetheweb** [comment]
IMHO I don't think a search page is really that good a secondary page to include to be honest. Think there are better candidates.

**Comment by rockeynebhwani** [comment]
> IMHO I don't think a search page is really that good a secondary page to include to be honest. Think there are better candidates.

May be not in general but for ECommerce, mostly we think in terms of 4 page Types (HomePage / Category Landing pages / Product Listings & Search pages / Product Detail Pages). I am discounting checkout pages for now as I don't think we are going to cover that.

Once you are on search page, you can look for LCP element and there is very high chances (I would say 90%+... some sites may have content search and other edge cases and hence 90% chance) and it will link to a product detail page. So, for eCommerce sites being picked up with these methods, you would have covered 3 most common page types... 

@bobbyshaw - Would you like to add anything from your Ecommerce background?

**Comment by bobbyshaw** [comment]
@rockeynebhwani I agree with the page types you will typically find but I'm not sure e-commerce warrants its own algorithm - particularly one that would need to be run synchronously: home, search, product, rather than: home, [secondary pages in parallel]

I would vote for something along the lines of the methods suggested above by @rviscomi and @patrickhulce. As these are more generically applicable and are almost certain to find a product listing page and with the secondary call to action or random link methods, we are also with a chance of finding a product detail page. 

In addition, I think in e-commerce, almost all technologies used on the site will be found from a home and product listing page.  There will be some that are only found on the product detail page but I would imagine the diminishing rate of return would not make it worth it based on the discussion above.

**Comment by paulcalvano** [comment]
I wonder if it would make sense to use Google search to find candidates for the second page.   For example, you could configure the programmable search API (https://developers.google.com/custom-search/v1/reference/rest) to query the entire web,  submit a google seach for `site:www.example.com`, and then extract the 2nd, 3rd, etc link from the search results. 

Here's a quick and dirty example where I added a handful of domain names to a list and queried the search API for the third result (ie, .items[2].link).   It's not perfect, but I wonder if this could be tweaked to give us something we can use.

```
pacalvan$ for i in $(cat test.txt); do link=$(curl -s  "Accept: application/json"  "https://customsearch.googleapis.com/customsearch/v1?key={API_KEY}&cx={SEARCH_ENGINE_ID}&q=site:$i" | jq -r '.items[2].link');  echo -e "$i\t$link"; done
www.amazon.com	https://www.amazon.com/blackfriday
www.google.com	https://www.google.com/partners/
www.youtube.com	https://www.youtube.com/channel/UCEIhICHOQOonjE6V0SLdrHQ
www.facebook.com	https://www.facebook.com/Rugbyrama/
www.yahoo.com	https://www.yahoo.com/subscriptions
www.netflix.com	https://www.netflix.com/pk/
www.microsoft.com	https://www.microsoft.com/en-sa
www.reddit.com	https://www.reddit.com/r/TheTrove/
www.instagram.com	https://www.instagram.com/paribucom/
www.ebay.com	https://www.ebay.com/deals
```

Any SEOs here familiar with this API, or similar ways of searching for popular pages?
